File: server/src/services/catchmentAnalysis.js
/**
 * Catchment Analysis Service
 * 
 * Delineates the catchment area (watershed) contributing to a selected pond site.
 * 
 * Method:
 * - Uses D8 flow direction grid
 * - BFS/flood-fill upstream from the pond (pour point)
 * - Collects all cells whose flow path leads to the pond
 * - Converts catchment cells to a polygon
 * - Calculates geodesic area using Turf.js
 * 
 * Limitations:
 * - Grid resolution limits polygon detail
 * - Interpolated DEM may create artifacts in flat areas
 * - D8 single-direction flow does not handle distributary flow
 */

const turf = require('@turf/turf');
const { gridToCoords, haversineDistance } = require('../utils/geometry');
const { DR, DC } = require('./terrainAnalysis');

/**
 * Delineate catchment area for a given pond location.
 * @param {Object} terrainModel - Output from buildTerrainModel
 * @param {Object} pondSite - Selected pond site with row, col
 * @returns {Object} Catchment polygon and area calculations
 */
function delineateCatchment(terrainModel, pondSite) {
  const {
    flowDirection, flowAccumulation, elevationGrid,
    nRows, nCols, bounds, cellSizeLng, cellSizeLat, cellSizeMeters
  } = terrainModel;

  const pourRow = pondSite.row;
  const pourCol = pondSite.col;

  // Step 1: Trace all upstream cells using BFS
  const catchmentCells = traceUpstream(
    flowDirection, nRows, nCols, pourRow, pourCol
  );

  console.log(`  Catchment cells: ${catchmentCells.size}`);

  if (catchmentCells.size === 0) {
    return {
      polygon: null,
      areaSquareMeters: 0,
      areaHectares: 0,
      areaSquareKilometers: 0
    };
  }

  // Step 2: Convert catchment cells to a polygon
  const polygon = catchmentCellsToPolygon(
    catchmentCells, nRows, nCols, bounds, cellSizeLng, cellSizeLat
  );

  // Step 3: Calculate area
  let areaSquareMeters = 0;
  if (polygon) {
    areaSquareMeters = turf.area(polygon);
  } else {
    // Fallback: estimate from cell count
    areaSquareMeters = catchmentCells.size * cellSizeMeters * cellSizeMeters;
  }

  const areaHectares = areaSquareMeters / 10000;
  const areaSquareKilometers = areaSquareMeters / 1000000;

  return {
    polygon: polygon ? polygon.geometry : null,
    areaSquareMeters: Math.round(areaSquareMeters * 100) / 100,
    areaHectares: Math.round(areaHectares * 10000) / 10000,
    areaSquareKilometers: Math.round(areaSquareKilometers * 10000) / 10000
  };
}

/**
 * Trace all cells upstream of a pour point using reverse flow direction.
 * @param {Array} flowDirection - 2D flow direction grid
 * @param {number} nRows
 * @param {number} nCols
 * @param {number} pourRow - Pour point row
 * @param {number} pourCol - Pour point column
 * @returns {Set} Set of "row,col" strings for catchment cells
 */
function traceUpstream(flowDirection, nRows, nCols, pourRow, pourCol) {
  const catchment = new Set();
  const queue = [`${pourRow},${pourCol}`];
  catchment.add(`${pourRow},${pourCol}`);

  while (queue.length > 0) {
    const current = queue.shift();
    const [row, col] = current.split(',').map(Number);

    // Check all 8 neighbors to see if they flow INTO this cell
    for (let d = 0; d < 8; d++) {
      const nr = row + DR[d];
      const nc = col + DC[d];

      if (nr < 0 || nr >= nRows || nc < 0 || nc >= nCols) continue;

      const key = `${nr},${nc}`;
      if (catchment.has(key)) continue;

      // The opposite direction: if neighbor flows in direction (d+4)%8, it flows to this cell
      const oppositeDir = (d + 4) % 8;
      if (flowDirection[nr][nc] === oppositeDir) {
        catchment.add(key);
        queue.push(key);
      }
    }
  }

  return catchment;
}

/**
 * Convert a set of catchment grid cells into a GeoJSON polygon.
 * Creates individual cell polygons and unions them.
 */
function catchmentCellsToPolygon(catchmentCells, nRows, nCols, bounds, cellSizeLng, cellSizeLat) {
  const cellPolygons = [];

  for (const key of catchmentCells) {
    const [row, col] = key.split(',').map(Number);

    // Cell corners
    const minLng = bounds.minLng + col * cellSizeLng;
    const maxLng = bounds.minLng + (col + 1) * cellSizeLng;
    const minLat = bounds.minLat + row * cellSizeLat;
    const maxLat = bounds.minLat + (row + 1) * cellSizeLat;

    cellPolygons.push(turf.bboxPolygon([minLng, minLat, maxLng, maxLat]));
  }

  if (cellPolygons.length === 0) return null;

  if (cellPolygons.length === 1) return cellPolygons[0];

  // Union all cell polygons
  try {
    // Use dissolve or iterative union
    let combined = cellPolygons[0];
    for (let i = 1; i < cellPolygons.length; i++) {
      try {
        const result = turf.union(
          turf.featureCollection([combined, cellPolygons[i]])
        );
        if (result) combined = result;
      } catch (e) {
        // Skip cells that fail to union (edge cases)
        continue;
      }
    }
    return combined;
  } catch (err) {
    // Fallback: use convex hull of all cell centers
    console.warn('  Union failed, using convex hull instead:', err.message);
    const centers = [];
    for (const key of catchmentCells) {
      const [row, col] = key.split(',').map(Number);
      const [lng, lat] = gridToCoords(row, col, bounds, cellSizeLng, cellSizeLat);
      centers.push(turf.point([lng, lat]));
    }
    return turf.convex(turf.featureCollection(centers));
  }
}

module.exports = { delineateCatchment };

File: python_server/services/catchment_analysis.py
"""
Catchment Analysis Service

Delineates the contributing catchment area upstream of the selected pond site.

Method:
  - Reverse BFS on the D8 flow direction grid from the pour point
  - Collect all cells whose flow path eventually reaches the pond
  - Build a GeoJSON polygon by unioning cell bounding boxes (via shapely)
  - Calculate geodesic area using the spherical excess formula

Identical algorithm to catchmentAnalysis.js.
"""

import math
from collections import deque
import numpy as np
from shapely.geometry import box, mapping
from shapely.ops import unary_union
from .terrain_analysis import DR, DC

EARTH_RADIUS_M = 6_371_000.0


def delineate_catchment(terrain_model: dict, pond_site: dict) -> dict:
    """
    Trace upstream catchment from pond_site and return polygon + area.
    """
    flow_dir = terrain_model['flow_direction']
    n_rows   = terrain_model['nRows']
    n_cols   = terrain_model['nCols']
    bounds   = terrain_model['bounds']
    csl      = terrain_model['cellSizeLng']
    cslat    = terrain_model['cellSizeLat']
    csm      = terrain_model['cellSizeMeters']

    pour_row = pond_site['row']
    pour_col = pond_site['col']

    # ── Trace upstream ─────────────────────────────────────────────────────
    catchment_cells = _trace_upstream(flow_dir, n_rows, n_cols, pour_row, pour_col)
    print(f"  Catchment cells: {len(catchment_cells)}")

    if not catchment_cells:
        return {'polygon': None, 'areaSquareMeters': 0.0, 'areaHectares': 0.0, 'areaSquareKilometers': 0.0}

    # ── Build polygon ──────────────────────────────────────────────────────
    polygon_geojson, area_m2 = _cells_to_polygon(catchment_cells, bounds, csl, cslat, csm)

    ha   = area_m2 / 10_000
    km2  = area_m2 / 1_000_000

    return {
        'polygon':           polygon_geojson,
        'areaSquareMeters':  round(area_m2 * 100) / 100,
        'areaHectares':      round(ha * 10_000) / 10_000,
        'areaSquareKilometers': round(km2 * 10_000) / 10_000,
    }


def _trace_upstream(flow_dir: np.ndarray, n_rows: int, n_cols: int,
                    pour_row: int, pour_col: int) -> set:
    """BFS reverse-trace: collect all cells that drain into (pour_row, pour_col)."""
    catchment = set()
    catchment.add((pour_row, pour_col))
    q = deque([(pour_row, pour_col)])

    while q:
        row, col = q.popleft()
        for d in range(8):
            nr = row + int(DR[d])
            nc = col + int(DC[d])
            if not (0 <= nr < n_rows and 0 <= nc < n_cols):
                continue
            if (nr, nc) in catchment:
                continue
            # Neighbour at direction d flows into (row,col) if its direction
            # points back (opposite = (d+4) % 8)
            opposite = (d + 4) % 8
            if int(flow_dir[nr, nc]) == opposite:
                catchment.add((nr, nc))
                q.append((nr, nc))

    return catchment


def _cells_to_polygon(cells: set, bounds: dict, csl: float, cslat: float, csm: float):
    """Union individual cell boxes into a single shapely polygon, compute area."""
    polys = []
    for (row, col) in cells:
        min_lng = bounds['minLng'] + col * csl
        max_lng = min_lng + csl
        min_lat = bounds['minLat'] + row * cslat
        max_lat = min_lat + cslat
        polys.append(box(min_lng, min_lat, max_lng, max_lat))

    try:
        merged = unary_union(polys)
    except Exception:
        # Fallback: convex hull
        from shapely.geometry import MultiPolygon
        merged = unary_union(polys).convex_hull

    area_m2 = _geodesic_area_m2(merged)
    return mapping(merged), area_m2


def _geodesic_area_m2(geom) -> float:
    """
    Approximate geodesic area using the spherical excess formula.
    Works on shapely Polygon or MultiPolygon.
    Falls back to cell-count estimate if geometry is unusual.
    """
    try:
        if geom.geom_type == 'MultiPolygon':
            return sum(_poly_area(p) for p in geom.geoms)
        return _poly_area(geom)
    except Exception:
        return 0.0


def _poly_area(poly) -> float:
    coords = list(poly.exterior.coords)
    area = 0.0
    for i in range(len(coords) - 1):
        lng1, lat1 = coords[i]
        lng2, lat2 = coords[i + 1]
        area += math.radians(lng2 - lng1) * (2 + math.sin(math.radians(lat1)) + math.sin(math.radians(lat2)))
    return abs(area) * EARTH_RADIUS_M ** 2 / 2
